In [120]:
import pandas as pd
import numpy as np
import os

In [121]:
os.chdir("C:\ml_using_python\Data\\")
#auto mpg data set has no header and they are space seperated
auto_mpg_df = pd.read_csv("auto-mpg.data", header=None, sep='\s+')

In [122]:
auto_mpg_df.shape

(398, 9)

In [123]:
auto_mpg_df.sample(8)

0  1      2      3       4     5   6  7                        8
154  15.0  6  250.0  72.00  3432.0  21.0  75  1          mercury monarch
211  16.5  6  168.0  120.0  3820.0  16.7  76  2       mercedes-benz 280s
65   14.0  8  351.0  153.0  4129.0  13.0  72  1         ford galaxie 500
96   13.0  8  360.0  175.0  3821.0  11.0  73  1  amc ambassador brougham
155  15.0  6  250.0  72.00  3158.0  19.5  75  1            ford maverick
179  22.0  4  121.0  98.00  2945.0  14.5  75  2              volvo 244dl
396  28.0  4  120.0  79.00  2625.0  18.6  82  1              ford ranger
313  28.0  4  151.0  90.00  2678.0  16.5  80  1       chevrolet citation

# give column / feature names

In [124]:
auto_mpg_df.columns=['mpg','cylinders','displacement','hp','weight','accelertation','year','origin','name']

In [125]:
auto_mpg_df.sample(8)

mpg  cylinders  displacement     hp  weight  accelertation  year  \
145  32.0          4          83.0  61.00  2003.0           19.0    74   
197  29.0          4          90.0  70.00  1937.0           14.2    76   
49   23.0          4         122.0  86.00  2220.0           14.0    71   
252  19.2          6         231.0  105.0  3535.0           19.2    78   
1    15.0          8         350.0  165.0  3693.0           11.5    70   
312  37.2          4          86.0  65.00  2019.0           16.4    80   
82   23.0          4         120.0  97.00  2506.0           14.5    72   
361  25.4          6         168.0  116.0  2900.0           12.6    81   

     origin                         name  
145       3                   datsun 710  
197       2                    vw rabbit  
49        1           mercury capri 2000  
252       1           pontiac phoenix lj  
1         1            buick skylark 320  
312       3                   datsun 310  
82        3  toyouta corona mark ii (sw)  
361       3              toyota cressida

In [126]:
auto_mpg_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 9 columns):
mpg              398 non-null float64
cylinders        398 non-null int64
displacement     398 non-null float64
hp               398 non-null object
weight           398 non-null float64
accelertation    398 non-null float64
year             398 non-null int64
origin           398 non-null int64
name             398 non-null object
dtypes: float64(4), int64(3), object(2)
memory usage: 28.1+ KB


__Let us find if there are any na or null values__

In [127]:
auto_mpg_df.isna().sum()

mpg              0
cylinders        0
displacement     0
hp               0
weight           0
accelertation    0
year             0
origin           0
name             0
dtype: int64

In [128]:
auto_mpg_df.isnull().sum()

mpg              0
cylinders        0
displacement     0
hp               0
weight           0
accelertation    0
year             0
origin           0
name             0
dtype: int64

__The feature which is numerical is shown as object, this may be because there are some non numerical values in this series__

In [129]:
feature_info = auto_mpg_df.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 9 columns):
mpg              398 non-null float64
cylinders        398 non-null int64
displacement     398 non-null float64
hp               398 non-null object
weight           398 non-null float64
accelertation    398 non-null float64
year             398 non-null int64
origin           398 non-null int64
name             398 non-null object
dtypes: float64(4), int64(3), object(2)
memory usage: 28.1+ KB


__By quick eyeballing  the hp feature there are some ? values present__

In [131]:
original_auto_df = auto_mpg_df
auto_mpg_df[auto_mpg_df.hp=='?']

mpg  cylinders  displacement hp  weight  accelertation  year  origin  \
32   25.0          4          98.0  ?  2046.0           19.0    71       1   
126  21.0          6         200.0  ?  2875.0           17.0    74       1   
330  40.9          4          85.0  ?  1835.0           17.3    80       2   
336  23.6          4         140.0  ?  2905.0           14.3    80       1   
354  34.5          4         100.0  ?  2320.0           15.8    81       2   
374  23.0          4         151.0  ?  3035.0           20.5    82       1   

                     name  
32             ford pinto  
126         ford maverick  
330  renault lecar deluxe  
336    ford mustang cobra  
354           renault 18i  
374        amc concord dl

__The above ? values in hp can be handled by usint to_numeric() method of pandas__

__errors = ignore, coerce, raise__

In [69]:
auto_mpg_df.hp = pd.to_numeric(auto_mpg_df.hp,errors='ignore')  #ignore will just return the same input ?

In [70]:
auto_mpg_df[auto_mpg_df.hp=='?']

mpg  cylinders  displacement hp  weight  accelertation  year  origin  \
32   25.0          4          98.0  ?  2046.0           19.0    71       1   
126  21.0          6         200.0  ?  2875.0           17.0    74       1   
330  40.9          4          85.0  ?  1835.0           17.3    80       2   
336  23.6          4         140.0  ?  2905.0           14.3    80       1   
354  34.5          4         100.0  ?  2320.0           15.8    81       2   
374  23.0          4         151.0  ?  3035.0           20.5    82       1   

                     name  
32             ford pinto  
126         ford maverick  
330  renault lecar deluxe  
336    ford mustang cobra  
354           renault 18i  
374        amc concord dl

In [71]:
auto_mpg_df.hp = pd.to_numeric(auto_mpg_df.hp,errors='raise')  #ignore will just return the same input ?

#This will raise error as shown below and ? will still sustain

ValueError: Unable to parse string "?" at position 32

In [72]:
auto_mpg_df[auto_mpg_df.hp=='?']

mpg  cylinders  displacement hp  weight  accelertation  year  origin  \
32   25.0          4          98.0  ?  2046.0           19.0    71       1   
126  21.0          6         200.0  ?  2875.0           17.0    74       1   
330  40.9          4          85.0  ?  1835.0           17.3    80       2   
336  23.6          4         140.0  ?  2905.0           14.3    80       1   
354  34.5          4         100.0  ?  2320.0           15.8    81       2   
374  23.0          4         151.0  ?  3035.0           20.5    82       1   

                     name  
32             ford pinto  
126         ford maverick  
330  renault lecar deluxe  
336    ford mustang cobra  
354           renault 18i  
374        amc concord dl

In [73]:
auto_mpg_df.hp = pd.to_numeric(auto_mpg_df.hp,errors='coerce')  #coerce will replace ? to NaN

In [74]:
auto_mpg_df[auto_mpg_df.hp=='?']

C:\Users\Durgaprasad.K\Continuum\miniconda3\lib\site-packages\pandas\core\ops.py:1649: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  result = method(y)


Empty DataFrame
Columns: [mpg, cylinders, displacement, hp, weight, accelertation, year, origin, name]
Index: []

In [75]:
auto_mpg_df.hp.isna().sum()

6

__Now we have converted ? to na we can handle them, by dropping them or replacing them with mean or median or mode__

In [76]:
auto_mpg_df.dropna(subset=['hp'], inplace=True)

In [77]:
auto_mpg_df.hp.isna().sum()

0

In [78]:
auto_mpg_df.shape

(392, 9)

__If we check the initial number of rows before dropping the na is 398 and after dropping it is 392, if there are more number of na and if relatively less dataset size, then dropping na will further decrease the data set__. 

__The imputation techniques should handle accordingly either by dropping or replacing the missing values with mean, median or mode. There can be cases where the replacement should be based on combo of mean,median and mode__

#  Filling the missing data with mean

In [132]:
original_auto_df.hp = pd.to_numeric(original_auto_df.hp, errors='coerce')

In [133]:
hp_mean = origina_auto_df.hp.mean()

In [134]:
original_auto_df.hp.fillna(hp_mean, inplace=True)

In [135]:
original_auto_df.hp.isna().sum()

0

In [136]:
origina_auto_df.shape

(398, 9)